src/dune/sub_dirs.ml

```diff
@@ -112,34 +112,69 @@ let eval (t : _ Status.Map.t) ~dirs =
                 dir
             ]))
 
-let decode =
+module Dir_map = struct
+  type per_dir =
+    { sexps : Dune_lang.Ast.t list
+    ; subdir_status : Predicate_lang.Glob.t Status.Map.t
+    }
+
+  type t =
+    { data : per_dir
+    ; nodes : t String.Map.t
+    }
+
+  let root t = t.data
+
+  let descend t (p : Path.Source.t) =
+    let components = Path.Source.explode p in
+    let rec loop t = function
+      | [] -> Some t.data
+      | x :: xs ->
+        let open Option.O in
+        let* next = String.Map.find t.nodes x in
+        loop next xs
+    in
+    loop t components
+end
+
+let strict_subdir field_name =
```

Comment:
This paor of functions were just moved outside of decode for clarity: `strict_subdir`, `strict_subdir_glob`.